### PR TITLE
Added info_type_transformations field in record_transformations to the google_data_loss_prevention_deidentify_template resource

### PR DESCRIPTION
**Pull request by abheda-crest**
Added `info_type_transformations` block in `record_transformations` field to the `google_data_loss_prevention_deidentify_template` resource.
fixes https://github.com/hashicorp/terraform-provider-google/issues/9836

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [x] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
dlp: added `info_type_transformations` block in the `record_transformations` field to `google_data_loss_prevention_deidentify_template` resource.
```

### CODE REVIEW

**Review by NickElliot — DISMISSED** [review]
From an initial passover, this is the main config change that needs to be made that I can see at this time. The extremely thorough test coverage looks good!

File: mmv1/products/dlp/DeidentifyTemplate.yaml [hunk 1]
```diff
@@ -771,10 +771,10 @@ properties:
                                           - :SUNDAY
                 - !ruby/object:Api::Type::NestedObject
                   name: primitiveTransformation
-                  required: true
                   description: |
```
[the remark this answers is not in the document]
Comment (abheda-crest) [reply mmv1/products/dlp/DeidentifyTemplate.yaml:774]: I agree with the point of adding `exactly_one_of` block but the `primitiveTransformations` and `infoTypeTranformations` subfields are a part of an array field [`fieldTransformations`](https://cloud.google.com/dlp/docs/reference/rest/v2/projects.deidentifyTemplates#fieldtransformation). Currently, `ConflictsWith`, `ExactlyOneOf` or `AtLeastOneOf` are set at the attribute level, which means we can't set it on a list if we have a list of objects, and each object has a list of attributes that's `ExactlyOneOf`. We don't know the length of the list and there is no way to differentiate object[0] from object[1] with what is conflicting. Hence, the use of the above mentioned functions when the attributes are part of array won't work. The issue addressing this limitation can be found [here](https://github.com/hashicorp/terraform-plugin-sdk/issues/470).

File: mmv1/products/dlp/DeidentifyTemplate.yaml [hunk 2]
```diff
@@ -1766,6 +1766,834 @@ properties:
                               description: |
                                 Words or phrases defining the dictionary. The dictionary must contain at least one phrase and every phrase must contain at least 2 characters that are letters or digits.
                               item_type: Api::Type::String
+                # infoTypeTransformations inside the recordTransformations
+                - !ruby/object:Api::Type::NestedObject
+                  name: infoTypeTransformations
+                  description: |
```
[the remark this answers is not in the document]
Comment (abheda-crest) [reply mmv1/products/dlp/DeidentifyTemplate.yaml:1772]: Same as above.

File: mmv1/products/dlp/DeidentifyTemplate.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (abheda-crest) [reply mmv1/products/dlp/DeidentifyTemplate.yaml:774]: I've added the note to provide only one of these field in the description though.

**Comment by abheda-crest** [comment]
@NickElliot I've resolved the merge conflict. Could you please review this PR?

**Comment by NickElliot** [comment]
LGTM! to @c2thorn for final review

**Review by c2thorn — CHANGES_REQUESTED** [review]
have concerns about two fields

File: mmv1/products/dlp/DeidentifyTemplate.yaml [hunk 3]
```diff
@@ -2104,6 +2104,834 @@ properties:
...
+                                            Name of the key. This is an arbitrary string used to differentiate different keys. A unique key is generated per name: two separate `TransientCryptoKey` protos share the same generated key if their names are the same. When the data crypto key is generated, this name is not used in any way (repeating the api call will result in a different key being generated).
+                                    - !ruby/object:Api::Type::NestedObject
+                                      name: unwrapped
+                                      description: |
+                                        Unwrapped crypto key. Using raw keys is prone to security risks due to accidentally leaking the key. Choose another type of key if possible. Only one of this, `transient` or `kms_wrapped` must be specified.
+                                      properties:
+                                        - !ruby/object:Api::Type::String
+                                          name: key
+                                          required: true
+                                          description: |
```
[the remark this answers is not in the document]
Comment (abheda-crest) [reply mmv1/products/dlp/DeidentifyTemplate.yaml]: Thanks for pointing that out. I've added `sensitive: true` to four `unwrapped.key` fields.

File: mmv1/products/dlp/DeidentifyTemplate.yaml [hunk 4]
```diff
@@ -2104,6 +2104,834 @@ properties:
...
+                            - !ruby/object:Api::Type::NestedObject
+                              name: characterMaskConfig
+                              description: |
+                                Partially mask a string by replacing a given number of characters with a fixed character. Masking can start from the beginning or end of the string. This can be used on data of any type (numbers, longs, and so on) and when de-identifying structured data we'll attempt to preserve the original data's type. (This allows you to take a long like 123 and modify it to a string like **3).
+                              properties:
+                                - !ruby/object:Api::Type::String
+                                  name: maskingCharacter
+                                  description: |
+                                    Character to use to mask the sensitive values—for example, * for an alphabetic string such as a name, or 0 for a numeric string
+                                    such as ZIP code or credit card number. This string must have a length of 1. If not supplied, this value defaults to * for
```
[the remark this answers is not in the document]
Comment (abheda-crest) [reply mmv1/products/dlp/DeidentifyTemplate.yaml]: I've added the description as per the [API docs](https://cloud.google.com/dlp/docs/reference/rest/v2/projects.deidentifyTemplates#charactermaskconfig). But, when I tried creating the resource without passing the `maskingCharacter`, the API didn't return its default value. Hence, I've not added `default_from_api: true`.
Following are the request and response of the API call:
```
POST /v2/projects/{project_id}/deidentifyTemplates?alt=json HTTP/1.1
---[ REQUEST ]---------------------------------------
{
 "deidentifyTemplate": {
  "deidentifyConfig": {
   "recordTransformations": {
    "fieldTransformations": [
     {
      "fields": [
       {
        "name": "unconditionally-char-masked-field"
       }
      ],
      "infoTypeTransformations": {
       "transformations": [
        {
         "infoTypes": [
          {
           "name": "EMAIL_ADDRESS"
          },
          {
           "name": "LAST_NAME"
          }
         ],
         "primitiveTransformation": {
          "characterMaskConfig": {
           "charactersToIgnore": [
            {
             "charactersToSkip": "-"
            }
           ],
           "numberToMask": 8,
           "reverseOrder": true
          },
          "redactConfig": null,
          "replaceWithInfoTypeConfig": null
         }
        }
       ]
      }
     }
    ],
    "recordSuppressions": [
     {
      "condition": {
       "expressions": {
        "conditions": {
         "conditions": [
          {
           "field": {
            "name": "field3"
           },
           "operator": "EQUAL_TO",
           "value": {
            "stringValue": "FOO-BAR"
           }
          },
          {
           "field": {
            "name": "field2"
           },
           "operator": "EQUAL_TO",
           "value": {
            "stringValue": "foobar"
           }
          },
          {
           "field": {
            "name": "field1"
           },
           "operator": "EQUAL_TO",
           "value": {
            "stringValue": "fizzbuzz"
           }
          }
         ]
        },
        "logicalOperator": "AND"
       }
      }
     }
    ]
   }
  },
  "description": "Description",
  "displayName": "Displayname"
 }
}
```
```
---[ RESPONSE ]--------------------------------------
{
  "name": "projects/{project_id}/deidentifyTemplates/1539292205506838093",
  "displayName": "Displayname",
  "description": "Description",
  "createTime": "2023-06-06T05:23:19.684737Z",
  "updateTime": "2023-06-06T05:23:19.684737Z",
  "deidentifyConfig": {
    "recordTransformations": {
      "fieldTransformations": [
        {
          "fields": [
            {
              "name": "unconditionally-char-masked-field"
            }
          ],
          "infoTypeTransformations": {
            "transformations": [
              {
                "infoTypes": [
                  {
                    "name": "EMAIL_ADDRESS"
                  },
                  {
                    "name": "LAST_NAME"
                  }
                ],
                "primitiveTransformation": {
                  "characterMaskConfig": {
                    "numberToMask": 8,
                    "reverseOrder": true,
                    "charactersToIgnore": [
                      {
                        "charactersToSkip": "-"
                      }
                    ]
                  }
                }
              }
            ]
          }
        }
      ],
      "recordSuppressions": [
        {
          "condition": {
            "expressions": {
              "logicalOperator": "AND",
              "conditions": {
                "conditions": [
                  {
                    "field": {
                      "name": "field3"
                    },
                    "operator": "EQUAL_TO",
                    "value": {
                      "stringValue": "FOO-BAR"
                    }
                  },
                  {
                    "field": {
                      "name": "field2"
                    },
                    "operator": "EQUAL_TO",
                    "value": {
                      "stringValue": "foobar"
                    }
                  },
                  {
                    "field": {
                      "name": "field1"
                    },
                    "operator": "EQUAL_TO",
                    "value": {
                      "stringValue": "fizzbuzz"
                    }
                  }
                ]
              }
            }
          }
        }
      ]
    }
  }
}
```

**Comment by c2thorn** [comment]
Running tests in our CI

**Review by c2thorn — APPROVED** [review]

**Pull request merged**